modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+            methodWriter.returnValue();
+            methodWriter.endMethod();
+        }
+
+        // Write all fields:
+        for (FieldNode irFieldNode : irClassNode.getFieldsNodes()) {
+            visit(irFieldNode, writeScope);
+        }
+
+        // Write all functions:
+        for (FunctionNode irFunctionNode : irClassNode.getFunctionsNodes()) {
+            visit(irFunctionNode, writeScope);
+        }
+
+        // End writing the class and store the generated bytes.
+        classVisitor.visitEnd();
+        irClassNode.setBytes(classWriter.getClassBytes());
+    }
+
+    @Override
+    public void visitFunction(FunctionNode irFunctionNode, WriteScope writeScope) {
+        int access = Opcodes.ACC_PUBLIC;
+
+        if (irFunctionNode.isStatic()) {
+            access |= Opcodes.ACC_STATIC;
+        }
+
+        if (irFunctionNode.hasVarArgs()) {
+            access |= Opcodes.ACC_VARARGS;
+        }
+
+        if (irFunctionNode.isSynthetic()) {
+            access |= Opcodes.ACC_SYNTHETIC;
+        }
+
+        Type asmReturnType = MethodWriter.getType(irFunctionNode.getReturnType());
+        List<Class<?>> typeParameters = irFunctionNode.getTypeParameters();
+        Type[] asmParameterTypes = new Type[typeParameters.size()];
+
+        for (int index = 0; index < asmParameterTypes.length; ++index) {
+            asmParameterTypes[index] = MethodWriter.getType(typeParameters.get(index));
+        }
+
+        Method method = new Method(irFunctionNode.getName(), asmReturnType, asmParameterTypes);
+
+        ClassWriter classWriter = writeScope.getClassWriter();
+        MethodWriter methodWriter = classWriter.newMethodWriter(access, method);
+        writeScope = writeScope.newMethodScope(methodWriter);
+
+        if (irFunctionNode.isStatic() == false) {
+            writeScope.defineInternalVariable(Object.class, "this");
+        }
+
+        List<String> parameterNames = irFunctionNode.getParameterNames();
+
+        for (int index = 0; index < typeParameters.size(); ++index) {
+            writeScope.defineVariable(typeParameters.get(index), parameterNames.get(index));
+        }
+
+        methodWriter.visitCode();
+
+        if (irFunctionNode.getMaxLoopCounter() > 0) {
+            // if there is infinite loop protection, we do this once:
+            // int #loop = settings.getMaxLoopCounter()
+
+            Variable loop = writeScope.defineInternalVariable(int.class, "loop");
+
+            methodWriter.push(irFunctionNode.getMaxLoopCounter());
+            methodWriter.visitVarInsn(Opcodes.ISTORE, loop.getSlot());
+        }
+
+        visit(irFunctionNode.getBlockNode(), writeScope.newBlockScope());
+
+        methodWriter.endMethod();
+    }
+
+    @Override
+    public void visitField(FieldNode irFieldNode, WriteScope writeScope) {
+        ClassWriter classWriter = writeScope.getClassWriter();
+        classWriter.getClassVisitor().visitField(
```

Comment:
Pull out some of these inline operations when calling this method, it's hard to read.

Reply:
Done.